layers.go

```diff
@@ -732,105 +731,102 @@ func (r *layerStore) Put(id string, parentLayer *Layer, names []string, mountLab
 	if mountLabel != "" {
 		label.ReserveLabel(mountLabel)
 	}
+
+	// Before actually creating the layer, make a persistent record of it with incompleteFlag,
+	// so that future processes have a chance to delete it.
+	layer := &Layer{
+		ID:           id,
+		Parent:       parent,
+		Names:        names,
+		MountLabel:   mountLabel,
+		Created:      time.Now().UTC(),
+		Flags:        make(map[string]interface{}),
+		UIDMap:       copyIDMap(moreOptions.UIDMap),
+		GIDMap:       copyIDMap(moreOptions.GIDMap),
+		BigDataNames: []string{},
+	}
+	r.layers = append(r.layers, layer)
+	r.idindex.Add(id)
+	r.byid[id] = layer
+	for _, name := range names {
+		r.byname[name] = layer
+	}
+	for flag, value := range flags {
+		layer.Flags[flag] = value
+	}
+	layer.Flags[incompleteFlag] = true
+
+	succeeded := false
+	cleanupFailureContext := ""
+	defer func() {
```

Comment:
Note that using `defer` means that we would try to delete the WIP layer even on a panic. That’s probably correct for some banal `nil` dereference, but might make things worse if the in-memory state were significantly corrupt.